### PR TITLE
Corrections and cleanup

### PR DESCRIPTION
**Pull request by flametail9**
Changes made based on review, plus final additions for 1d and 2d matrices.

### CODE REVIEW

File: projects/Mat_Test/Mat_test.cpp [hunk 1]
```diff
@@ -51,17 +55,82 @@ int main (){
     fprintf(outFile, "m.copy(newMat)\n");
     newMat.print(outFile);
 
-    output = m.roi(1,3);
-    fprintf(outFile, "m.roi(1,3)\n");
+    output = m.roi(0,1,1,3);
+    fprintf(outFile, "m.roi(0,1,1,3)\n");
     output.print(outFile);
```
Comment (anuranbaka) [inline projects/Mat_Test/Mat_test.cpp:60]: add 2 arg test and 1 arg test and 0 arg test

File: projects/Mat_Test/Mat_test.cpp [hunk 2]
```diff
@@ -51,17 +55,82 @@ int main (){
...
+    fprintf(outFile, "4x5 matrix of ones\n");
+    output.print(outFile);
+    
+    output = Mat<>::eye(6);
+    fprintf(outFile, "6x6 identity matrix\n");
+    output.print(outFile);
+
+    output = Mat<>::eye(4,7,1);
+    fprintf(outFile, "4x7 identity matrix with diagonal at 1\n");
+    output.print(outFile);
```
Comment (anuranbaka) [inline projects/Mat_Test/Mat_test.cpp:133]: add test  for start at negative cols

File: projects/matlib/Mat.h [hunk 3]
```diff
@@ -36,58 +42,110 @@ class Mat {
...
         size_type columns() const{
-            return this->dims[1];
+            return this->dims[ndims - 1];
         }
-        size_type size() const{
-            return rows()*columns();
+        bool inbounds(size_type a, size_type b){
+            if(a >= 0 && a < columns() && b >= 0 && b < rows()) return true;
+            else return false;
         }
```
Comment (anuranbaka) [inline projects/matlib/Mat.h:70]: 1D version

File: projects/matlib/Mat.h [hunk 4]
```diff
@@ -36,58 +42,110 @@ class Mat {
...
-        Mat(size_type a = 1, size_type b = 1){
+        bool isContiguous(){
+            if(strides[0] != 1) return false;
+            for(int i = 1; i < ndims; i++){
+                strides[i] != dims[i];
```
Comment (anuranbaka) [inline projects/matlib/Mat.h:74]: do something, and chain or not implemented error for higher dims

File: projects/matlib/Mat.h [hunk 5]
```diff
@@ -36,58 +42,110 @@ class Mat {
...
-            memory = new Type[a*b];
+            ndims = 1;
+            dims = new size_type[ndims];
+            dims[0] = a;
+            strides = new size_type[ndims];
+            strides[0] = 1;
+            memory = new Type[a];
             data = memory;
+        }
+        Mat(size_type a, size_type b){
```
Comment (anuranbaka) [inline projects/matlib/Mat.h:88]: signature
Mat(Type*, int64_t*)

usage:
double a[4] = {1,2,3,4};
int64_t counter=1;
Mat wrapper = Mat(a, &counter);

Comment (anuranbaka) [reply projects/matlib/Mat.h:88]: additional signature
Mat(Type*)
that automatically makes a reference counter that starts at 1

File: projects/matlib/Mat.h [hunk 6]
```diff
@@ -99,12 +157,16 @@ class Mat {
...
             delete []strides;
         }
+        Type& operator() (size_type a){
+            return data[a*strides[0]];
+        }
         Type& operator() (size_type a, size_type b){
-            errorCheck(a < 0 || b < 0 || a + b > size(),"Element access outside matrix scope\n");
             return data[a*strides[1] + b*strides[0]];
         }
+        Type& operator() (size_type a) const{
```
Comment (anuranbaka) [inline projects/matlib/Mat.h:166]: return const

File: projects/matlib/Mat.h [hunk 7]
```diff
@@ -3,6 +3,12 @@
 #include <initializer_list>
 using namespace std;
 
+template <class Type>
+Type Add(Type a, Type b){ return a + b; };
```
Comment (anuranbaka) [inline projects/matlib/Mat.h:7]: add inline keyword

File: projects/matlib/Mat.h [hunk 8]
```diff
@@ -3,6 +3,12 @@
...
+
+template <class Type>
+Type Subtract(Type a, Type b){ return a - b; };
```
Comment (anuranbaka) [inline projects/matlib/Mat.h:10]: add inline keyword

File: projects/matlib/Mat.h [hunk 9]
```diff
@@ -269,36 +371,108 @@ class Mat {
...
+        static Mat eye(size_type a){
+            Mat result(a,a);
+            for(size_type i = 0; i < a; i++){
+                for(size_type j = 0; j < a; j++){
+                    if(i == j) result(i,j) = 1;
+                    else result(i,j) = 0;
+                }
+            }
+            return result;
+        }static Mat eye(size_type a, size_type b, size_type k = 0){
```
Comment (anuranbaka) [inline projects/matlib/Mat.h:467]: endl

File: projects/matlib/Mat.h (hunk 5 above)
Comment (flametail9) [reply projects/matlib/Mat.h:92]: What do if original container deletes data?

File: projects/matlib/Mat.h [hunk 10]
```diff
@@ -43,7 +43,7 @@ class Mat {
         }
         iterator end(){
             if(ndims == 2){
-                return iterator(*this, strides[0]*columns() + strides[1]*rows());
+                return iterator(*this, strides[0]*columns()+strides[1]*rows());
```
Comment (anuranbaka) [inline projects/matlib/Mat.h:46]: usu space around polynomial +

File: projects/matlib/Mat.h [hunk 11]
```diff
@@ -149,8 +156,15 @@ class Mat {
         }
         ~Mat(){
```
Comment (anuranbaka) [inline projects/matlib/Mat.h:157]: reorder

File: projects/matlib/Mat.h [hunk 12]
```diff
@@ -174,10 +188,11 @@ class Mat {
...
         Mat& operator= (const Mat &b){
             (*refCount)--;
-            errorCheck(*refCount < 0,"Reference counter is negative somehow\n");
+            errorCheck(*refCount < 0,
+                "Reference counter is negative somehow\n");
             if(*refCount == 0){
                 delete refCount;
-                delete[] memory;
+                if(memory != nullptr) delete[] memory;
             }
```
Comment (anuranbaka) [inline projects/matlib/Mat.h:196]: call destructor

File: projects/matlib/Mat.h [hunk 13]
```diff
@@ -214,24 +229,35 @@ class Mat {
...
+                        (*f)(operator()(leftRow, leftCol),
+                            b(rightRow, rightCol));
                 }
             }
             else{
                 Mat<Type> temp(x[0]);
                 result = temp;
                 for(size_type i = 0; i < result.size(); i++){
-                    result(i) = (*f)(operator()(i%columns()) , b(i%b.columns()));
+                    result(i) = (*f)(operator()(i%columns()),b(i%b.columns()));
```
Comment (anuranbaka) [inline projects/matlib/Mat.h:260]: put back

File: projects/matlib/Mat.h [hunk 14]
```diff
@@ -306,10 +344,14 @@ class Mat {
...
-            errorCheck(memory == dest.memory, "Source and destination matrix share same backing data\n");
-            errorCheck(data == dest.data, "TODO: call other transpose function\n");
+            errorCheck(ndims != 2,
+                "transpose may only be used on 2d matrix");
+            errorCheck(dims[0] != dest.dims[1] || dims[1] != dims[0],
+                "Matrix size mismatch");
+            errorCheck(memory == dest.memory,
+                "Source and destination matrix share same backing data");
+            errorCheck(data == dest.data,
+                "TODO: call other transpose function");
```
Comment (anuranbaka) [inline projects/matlib/Mat.h:354]: implement

Comment (anuranbaka) [reply projects/matlib/Mat.h:354]: move above memory check

File: projects/matlib/Mat.h [hunk 15]
```diff
@@ -532,5 +696,4 @@ class MatIter{
         Type & operator*(){
             return matrix.data[position];
         }
-
-};
+};
```
Comment (anuranbaka) [inline projects/matlib/Mat.h:699]: newline
